src/Type/Constant/ConstantArrayType.php

```diff
@@ -388,6 +388,25 @@ public function isSuperTypeOf(Type $type): TrinaryLogic
 				$results[] = $isValueSuperType;
 			}
 
+			if (
+				$this->isList->yes()
+				&& $type->isList->yes()
+				&& count($this->keyTypes) !== count($type->keyTypes)
+				&& count($type->optionalKeys) === 0
+			) {
+				$keepSeparate = true;
+				foreach ($this->valueTypes as $valueType) {
+					if ($valueType->isConstantValue()->yes()) {
+						$keepSeparate = false;
+						break;
+					}
+				}
```

Comment:
we somehow need to identify which constant arrays we want to keep separte (in our case tagged unions), without running into "is always false" comparison false positives like 

> list<array{string, 1|2|3|4|5|6|7}> and list<array{string, 1|2|3|4|5|6|7, int}> will always evaluate to false

I think my approach is wrong, but I have no better idea atm

Reply:
Any change to isSuperTypeOf logic needs tests in TypeCombinatorTest::dataUnion and dataIntersect. So please express there what you're trying to achieve, only when reviewing that I can decide if it makes sense or not.

Reply:
I could not figure out a dataUnion and dataIntersect test.

but what made sense to me was a test like
subtract `array{mixed}` from `array{mixed, string|null, mixed}|array{mixed}` should lead to `array{mixed, string|null, mixed}`